src/bpftrace.cpp

```diff
@@ -2129,23 +2157,40 @@ bool BPFtrace::is_aslr_enabled(int pid)
   return false;
 }
 
-std::string BPFtrace::resolve_usym(uintptr_t addr, int pid, bool show_offset, bool show_module)
+std::string BPFtrace::resolve_usym(uintptr_t addr,
+                                   int pid,
+                                   int probe_id,
+                                   bool show_offset,
+                                   bool show_module)
 {
   struct bcc_symbol usym;
   std::ostringstream symbol;
   void *psyms = nullptr;
-  struct bcc_symbol_option symopts;
-
-  memset(&symopts, 0, sizeof(symopts));
-  symopts.use_debug_file = 1;
-  symopts.check_debug_file_crc = 1;
-  symopts.use_symbol_type = BCC_SYM_ALL_TYPES;
 
   if (resolve_user_symbols_)
   {
-    if (cache_user_symbols_)
+    std::string pid_exe = get_pid_exe(pid);
+    if (pid_exe.empty() && probe_id != -1)
+    {
+      // get program name from probe
+      auto probe_full = resolve_probe(probe_id);
+      size_t start = probe_full.find(':') + 1;
+      size_t end = probe_full.find(':', start);
+      pid_exe = probe_full.substr(start, end - start);
+    }
+    if (cache_user_symbols_per_program_)
     {
-      std::string pid_exe = get_pid_exe(pid);
+      if (!pid_exe.empty() && !show_offset && !show_module)
+      {
+        // try to resolve symbol directly from binary
+        std::map<uint64_t, std::string> &symbol_table =
+            symbol_table_cache_.find(pid_exe) != symbol_table_cache_.end()
+                ? symbol_table_cache_[pid_exe]
+                : (symbol_table_cache_[pid_exe] = get_symbol_table_for_elf(
+                       pid_exe));
```

Comment:
The issue with using insert is that under eager evaluation this always calls `get_symbol_table_for_elf`, making it less efficient.

Reply:
Ah, right. This looks ugly, but seems like we don't have much of a better option.
Perhaps this could be an alternative, but I'm not strongly against leaving the original one:
```suggestion
        if (symbol_table_cache_.find(pid_exe) == symbol_table_cache_.end())
            symbol_table_cache_[pid_exe] = get_symbol_table_for_elf(pid_exe)
        auto &symbol_table = symbol_table_cache_[pid_exe];
```